glean-core/src/database/mod.rs

```diff
@@ -147,6 +165,49 @@ impl Database {
         }
     }
 
+    /// Checks if the 'user' lifetime metric is expired and, if so, removes it.
+    ///
+    /// ## Arguments
+    ///
+    /// * `storage_name`: The storage name to iterate over.
+    /// * `metric_id_slice`: The metric id to check.
+    /// ## Return value
+    ///
+    /// Returns a `true` if the metric is expired and should not be reported,
+    /// `false` otherwise.
+    ///
+    /// ## Panics
+    ///
+    /// This function will **not** panic on database errors.
+    fn check_and_remove_expired_user_lifetime(&self, storage_name: &str, metric_id_slice: &[u8]) -> bool {
+        let metric_name_str = match str::from_utf8(metric_id_slice) {
+            Ok(metric_name) => metric_name.to_string(),
+            _ => return false,
+        };
+
+        match USER_LIFETIME_EXPIRATION_MAP.get(metric_name_str.as_str()) {
+            Some(&expiration) => 
+            {
+                let is_expired = *STARTUP_LOCAL_DATE > expiration;
+                if is_expired {
+                    // This should be fine to delete now, even though we may be iterating
+                    // through the database entries.
+                    if let Err(e) = self.remove_single_metric(Lifetime::User, storage_name, &metric_name_str) {
```

Comment:
@badboy this seems to be safe, no test is complaining. I was a bit surprised it worked, but then I realized that LMDB can probably handle reading vs writing while iterating. A different strategy would be to move this code to `iter_store_from`, store the expired ids in a vector and then remove them outside of the `while`.

Reply:
@badboy any comment on this?

Reply:
hm ... I need to re-read that code and also rkv write vs read mode, just to make sure.
I'll get back to you.